src/servers/src/http/json_result.rs

```diff
@@ -0,0 +1,135 @@
+pub struct JsonResponse {
+    output: Vec<GreptimeQueryOutput>,
+    execution_time_ms: u64,
+}
+
+impl JsonResponse {
+    pub async fn from_output(outputs: Vec<crate::error::Result<Output>>) -> HttpResponse {
+        match handler::from_output(outputs).await {
+            Err(err) => HttpResponse::Error(err),
+            Ok((output, _)) => {
+                if output.len() > 1 {
+                    HttpResponse::Error(ErrorResponse::from_error_message(
+                        StatusCode::InvalidArguments,
+                        "cannot output multi-statements result in json format".to_string(),
+                    ))
+                } else {
+                    HttpResponse::Json(JsonResponse {
+                        output,
+                        execution_time_ms: 0,
+                    })
+                }
+            }
+        }
+    }
+
+    pub fn output(&self) -> &[GreptimeQueryOutput] {
+        &self.output
+    }
+
+    pub fn with_execution_time(mut self, execution_time: u64) -> Self {
+        self.execution_time_ms = execution_time;
+        self
+    }
+
+    pub fn execution_time_ms(&self) -> u64 {
+        self.execution_time_ms
+    }
+
+    pub fn with_limit(mut self, limit: usize) -> Self {
+        self.output = process_with_limit(self.output, limit);
+        self
+    }
+}
+
+impl IntoResponse for JsonResponse {
+    fn into_response(mut self) -> Response {
+        debug_assert!(
+            self.output.len() <= 1,
+            "self.output has extra elements: {}",
+            self.output.len()
+        );
+
+        let execution_time = self.execution_time_ms;
+        let payload = match self.output.pop() {
+            None => String::default(),
+            Some(GreptimeQueryOutput::AffectedRows(n)) => json!({
+                "data": [
+                    {
+                        "affectedrows": n
+                    },
+                ],
+                "execution_time_ms": execution_time,
+            })
+            .to_string(),
+
+            Some(GreptimeQueryOutput::Records(records)) => {
+                let mut data = Vec::new();
+                let schema = records.schema();
+
+                for row in records.rows.iter() {
+                    let mut row_map = Map::new();
+                    for (i, col) in schema.column_schemas.iter().enumerate() {
+                        row_map.insert(col.name.clone(), row[i].clone());
+                    }
+                    data.push(row_map);
+                }
+
+                json!({
+                    "data": data,
+                    "execution_time_ms": execution_time,
+                })
+                .to_string()
+            }
+        };
+
+        let mut resp = (
+            [(
+                header::CONTENT_TYPE,
+                HeaderValue::from_static(mime::APPLICATION_JSON.as_ref()),
+            )],
+            payload,
+        )
+            .into_response();
+        resp.headers_mut().insert(
```

Comment:
Why not move these headers above into the construction of `resp`? Then we don't need to `insert` them.